sstables/trie: add BTI index readers and writers

michoecho opened the pull request:
This is yet another part in the BTI index project.

Overarching issue: https://github.com/scylladb/scylladb/issues/19191
Previous part: https://github.com/scylladb/scylladb/pull/25506/
Next part: plugging the BTI index readers and writers into sstable readers and writers.

The new code added in this PR isn't used outside of tests yet, but it's posted as a separate PR for reviewability.

This series implements, on top of the key translation logic, and abstract trie writing and traversal logic, a writer and a reader of sstable index files (which map primary keys to positions in Data.db), as described in https://github.com/apache/cassandra/blob/f16fb6765b8a3ff8f49accf61c908791520c0d6e/src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormat.md.  

Caveats:
1. I think the added test has reasonable coverage, but that depends on running it multiple times. (Though it shouldn't need more than a few runs to catch any bug it covers). It's somewhat awkward as a test meant for running in CI, it's better as something you run many times after a relevant change.
2. These readers and writers are intended to be compatible with Cassandra, but I did *NOT* do any compatibility testing. The writers and readers added here have only been tested against each other, not against Cassandra's readers and writers.
3. This didn't undergo any proper benchmarking and optimization work. I was doing some measurements in the past, but everything was rewritten so much since then that the my old measurements are effectively invalidated. Frankly I have no idea what the performance of all this branchy-branchy logic is now.

No backports needed, new functionality.

avikivity commented:
> 2. These readers and writers are intended to be compatible with Cassandra, but I did _NOT_ do any compatibility testing. The writers and readers added here have only been tested against each other, not against Cassandra's readers and writers.

We will need compatibility testing - reading Cassandra sstables is sufficient, and having Cassandra read our own. The latter can't be done in a unit test without importing the Cassandra code base, so we'll do it as an external test.

I'm okay with deferring this to the next step if you think that's where the read compatibility tests should go.

avikivity commented on test/boost/bti_key_translation_test.cc:
```diff
@@ -303,7 +303,9 @@ BOOST_AUTO_TEST_CASE(test_lcb_mismatch) {
-                    co_yield std::ranges::elements_of(r);
+                    for (auto e : r) {
+                        co_yield e;
+                    }
```
What's wrong with the original code? It looks like it's correct.

I don't mind the change, just curious.

michoecho replied:
Type mismatch. `std::ranges::elements_of(r)` generates references, not values. But this generator wants to return values, not references.

michoecho replied:
(And it makes perfect sense.)

avikivity replied:
https://lists.isocpp.org/std-proposals/2025/08/15009.php

avikivity commented on sstables/trie/bti_key_translation.cc:
```diff
@@ -133,8 +133,22 @@ static bound_weight convert_bound_to_bound_weight(sstables::bound_kind_m b) {
 
 lazy_comparable_bytes_from_clustering_position::lazy_comparable_bytes_from_clustering_position(const schema& s, position_in_partition_view pipv)
     : _ckp(pipv.has_key() ? std::optional<clustering_key_prefix>(pipv.key()) : std::optional<clustering_key_prefix>())
```
> are never passed sstable indexes

passed to?

avikivity commented on sstables/trie/trie_writer.hh:
```diff
@@ -35,6 +35,7 @@
 #include <seastar/util/log.hh>
 #include <map>
```
> Note: this might be overengineered. Maybe we should force the
user to linearize the key. Maybe caring so much about fragmented
keys as first-class citizens is the wrong thing to do.

1. keys are not fragmented as long as they're outside lsa (even stored as bytes, not managed_bytes)
2. I'd like to eventually convert to comparably_bytes early (at the coordinator)

But we can undo these later.

michoecho replied:
Yes, undoing the fragmentation is always easier than introducing it.

avikivity commented on sstables/trie/bti_index.hh:
```diff
@@ -0,0 +1,83 @@
+    void add(
+        const schema& s,
+        const sstables::clustering_info& first_ck,
+        const sstables::clustering_info& last_ck,
```
For the caller, what's the best strategy here? Call every 4k of data? Why not even less?

avikivity commented on configure.py:
```diff
@@ -565,6 +565,7 @@ def find_ninja():
     'test/boost/token_metadata_test',
     'test/boost/top_k_test',
```
> and before releases. Unfortunately we don't have a precedent for
something like that.

@mykaul let's start work on an intermediate category of tests like this one. We already have tests that benefit from longer runtime but don't merit extending CI time for them.

avikivity commented:
Looks very good.

I'm convinced we need Cassandra interoperability tests, there are too many details here.

michoecho commented on test/boost/bti_index_test.cc:
```diff
@@ -0,0 +1,1026 @@
+    // Should be at most as big as clustering_universe_size / 2,
+    // otherwise there isn't enough clustering positions to fill the blocks.
+    int max_clustering_blocks = clustering_universe_size / 2;
+};
```
> Will this test partition keys that span a page?

No.

michoecho replied:
Added a separate test for this, and it did find a bug. (Though arguably it's a bug in the pre-existing `processing_result_generator` rather than in my `row_index_header_parser`).

michoecho commented on sstables/trie/bti_node_reader.hh:
```diff
@@ -34,7 +34,21 @@ node_traverse_sidemost_result bti_walk_down_rightmost_path(int64_t pos, const_by
+            return *this;
+        }
+    };
+
```
> Wait, that probably doesn't work. Likely most pages in cache have a zero use_count.

They do. Maybe it would be possible to implement `cached_page` on top of `enable_lw_shared_from_this`, but it would be awkward enough that I'm not sure it's worth trying.

> It's better to move copying support to cached_page, but I don't mind to defer that.

It doesn't matter much, does it.

michoecho commented on sstables/trie/bti_key_translation.cc:
```diff
@@ -133,8 +133,22 @@ static bound_weight convert_bound_to_bound_weight(sstables::bound_kind_m b) {
+        _terminator = std::byte(0xff);
+    } else {
+        _terminator = bound_weight_to_terminator(pipv.get_bound_weight());
+    }
```
> Can we somehow on_internal_error during writing?

Here? We would have to add some `bool are_we_writing` just to implement this check here. But I think that would be overkill.

There's nothing terribly wrong with writing non-clustering positions with this bespoke encoding somewhere. It's just that the row index isn't prepared for that. So it's the row index writer's job not to write non-clustering positions. And that's not an easy mistake to make, because the row index writer doesn't even see non-clustering positions. It's only fed `clustering_key_prefix` objects.

michoecho commented:
> We will need compatibility testing - reading Cassandra sstables is sufficient, and having Cassandra read our own. The latter can't be done in a unit test without importing the Cassandra code base, so we'll do it as an external test.

@avikivity To clarify: do you want me to implement a Java test in the Cassandra codebase that will try to do some detailed checks on Scylla-generated index files (since you're talking about "importing the cassandra code base"), or just a black box test which uses Cassandra to read those files through regular CQL queries?

avikivity commented:
> > We will need compatibility testing - reading Cassandra sstables is sufficient, and having Cassandra read our own. The latter can't be done in a unit test without importing the Cassandra code base, so we'll do it as an external test.
> 
> @avikivity To clarify: do you want me to implement a Java test in the Cassandra codebase that will try to do some detailed checks on Scylla-generated index files (since you're talking about "importing the cassandra code base"), or just a black box test which uses Cassandra to read those files through regular CQL queries?

I think a black box test is okay (but not in this repository, maybe we'll open a new one for this).

We need two directions: prove that we can read Cassandra indexes, and prove that Cassandra can read out indexes.

michoecho commented:
> btw, it's not as simple as issuing queries, since the data read phase can cover errors in indexing. If the index points to early, the read might still succeed.

That's why I'm asking. With a black box (i.e. via CQL) it's impossible to prove, for example, that Cassandra doesn't ignore the Scylla-generated index completely and opts to scan the entire Data file instead. You can't really stress the index layer via CQL, you can only check that queries succeed and return correct results.

On the other hand, writing white-box tests (in Java) that would test each layer separately and well (key encoding, node encoding, trie iteration logic, etc.) would take weeks.

I've spent several hours thinking about this, and my plan so far is:
1. Generate a BIG sstable (with a few million entries, which hopefully also exercises various node types and sizes) with a Scylla-based test executable. Remember the positions of each fragment.
2. `nodetool import` the sstable into Cassandra. Let it upgrade it to BTI.
3. Load the Cassandra-generated BTI index into the Scylla test executable. Iterate over all entries in the index, get their Data positions, pick the corresponding fragments, and generate a Scylla-BTI index with the same entries.
4. Iterate over both indexes and check that their return the same positions and tombstones.
5. Run index queries on the Cassandra index with a Scylla reader for all positions present in the index, check that that the results are as expected.
6. In Cassandra, clone the table with the Cassandra-BTI index, in the new table swap the BTI index files for Scylla-generated ones, run various CQL queries on both tables and check that they return the same results.

That's the best idea I could come up with without getting into Java.

nyh commented:
> > btw, it's not as simple as issuing queries, since the data read phase can cover errors in indexing. If the index points to early, the read might still succeed.
> 
> That's why I'm asking. With a black box (i.e. via CQL) it's impossible to prove, for example, that Cassandra doesn't ignore the Scylla-generated index completely and opt to scan the entire Data file instead.

You could use strace or fuse or whatever to learn whether Cassandra actually read the relevant parts of the index, but I agree this can be a bit funky.

Another funky (perhaps slightly less funky) thing you could do is to deliberately corrupt the data file **outside** the byte range you expect Cassandra to read, and then you'll know Cassandra really used the index and didn't try to read the data file.

> You can't really stress the index layer via CQL, you can only check that queries succeed and return correct results.
> 
> On the other hand, writing white-box tests (in mixed Java/C++) that would test each layer separately and well (key encoding, node encoding, trie iteration logic, etc.) would take weeks.

The biggest problem I see with such "white box" testing is that they are great for checking your development, but they make for bad regression testing for years to come: If a year from now, anything changes in either Cassandra's or Scylla's implementation, many of these tests will just break and the busy developers of a year from now will probably just skip them instead of trying to figure out how to do major changes to the test to make it pass.

avikivity commented:
If Cassandra tracing is as detailed as ours, we could use the tracing output.

michoecho commented:
v2:
- Added a Cassandra compatibility test, under `test/manual`. See `bti_cassandra_compatibility_test.cc` and `bti_cassandra_compatibility.py`.
- Fixed several incompatibilities revealed by the test:
  - Partitions.db contains a 1-byte hash of the partition key, to minimize the number of false positives. v1 of this PR was using the last byte of the token for this purpose. But it turns out that Cassandra uses the last byte of the whole 16-byte murmur hash, not the last byte of the token. (The last byte of the token is byte 7 of the murmur hash). This was fixed, but introduced a new FIXME: in v2, we compute the murmur hash inside the index readers and writers, which a waste because the callers of the index already have to compute it for other purposes. It would be better to pass the hash down from the caller. But a better place for that is in another PR, where we hook up the index to the callers.
  - In partition headers in Rows.db, v1 of this PR stored the root offset (i.e. offset between the header and the root of the row trie) as an unsigned vint. Cassandra stores it as a signed vint (which is always negative in practice).
  - In partition headers in Rows.db, v1 of this PR was missing the `number of clustering blocks` unsigned vint field. This was fixed.(Note: as far as I can see, this field is useless).
  - In the payloads in Rows.db, Cassandra stores the offset of the row within the partition as a signed vint. v1 of this PR was using an unsigned vint. This was fixed. (Note: unsigned vint makes more sense, because this value can't be negative. Using a signed vint wastes 1 bit).
  - 
- Added a test (`test_read_row_index_header`) for parsing Rows.db partition headers which cross page boundaries.
  - This found a bug — if the parsing requires an await, and the `input_stream` extends beyond the header, then `continuous_data_consumer` would call the `processing_result_generator` coroutine again after it has already returned, triggering a nullptr dereference. I fixed this by adding an explicit `co_yield data_consumer::proceed::no;` before both return points in `row_index_header_parser::do_process_state()`. (Although this is really a bug in `processing_result_generator` — if the coroutine returned, then obviously `processing_result_generator::generate()` should return `proceed::no`. Currently it returns whatever the result of the last `co_yield` before the return was.)
- Fixed some bugs (not compatibility bugs, just bugs) revealed by running the randomized compatibility test:
  - The logic for choosing the trie key for `partition_end` was wrong. This key is chosen arbitrarily — it only needs to be greater than the last key in the partition. But, following Cassandra, we choose it by trimming the last clustering key to the actually-inserted length of the preceding key, and "nudging" it by incrementing the last byte. Except somehow I forgot that this only works when the last byte isn't `0xff`. If that's the case, we can't trim the last key to this position — we have to choose some non-`0xff` byte. See `bti_row_index_writer.cc::nudge()`, also following Cassandra. (Note: this is something that shows up in random tests easily when key values are randomized, but it didn't show up in my `bti_index_test::test_exhaustive` because I deliberately chose sequential integers as keys, for debuggability, so there were no 0xff bytes in keys. In v2, I changed the last of these sequential integers to `0xff`, so that the test catches this bug).
  - `max_offset_from_child` (a function introduced in a previous PR) was wrongly handling a special case, breaking trie writer's size calculations. See `sstables/trie: fix a special case in max_offset_from_child` for the details.

michoecho commented on sstables/trie/bti_index.hh:
```diff
@@ -0,0 +1,83 @@
+    // The trie will be written to the given file writer.
+    // Note: the file doesn't have to be empty,
+    // and it can be extended later.
+    bti_row_index_writer(sstables::file_writer&);
```
Done.

avikivity commented on test/manual/bti_cassandra_compatibility_test.py:
```diff
@@ -0,0 +1,621 @@
+    await clean_gather(*[cancel_process(p, timeout=60) for p in procs])
+    # The cassandra image chowns /var/lib/cassandra to the `cassandra` user
+    # on startup, which means that *we* (the caller of this entire Python script)
+    # lose permissions to it. That's uncool, so we chown it back.
```
That's really bad. But I see

```sh
# allow the container to be started with `--user`
if [ "$1" = 'cassandra' -a "$(id -u)" = '0' ]; then
	find "$CASSANDRA_CONF" /var/lib/cassandra /var/log/cassandra \
		\! -user cassandra -exec chown cassandra '{}' +
	exec gosu cassandra "$BASH_SOURCE" "$@"
fi
```


can we break the condition here? I guess not, uid would be zero in dbuild and "$1" has to be cassandra to execute anything.

Maybe bind-mount something to /etc/passwd such that cassandra has uid 0?

I'm trying to avoid abnormal termination -> randomly owned files in my home directory.

michoecho replied:
> can we break the condition here?

I tried to find some way to hack around the condition, and I didn't see one.

> I'm trying to avoid abnormal termination -> randomly owned files in my home directory.

I tried to make it so that Python exceptions and SIGINT won't leave randomly-owned files. But the cleanup won't work SIGKILL, power loss, etc.

> Maybe bind-mount something to /etc/passwd such that cassandra has uid 0?

Is that even a thing? Can you give a named user (i.e. non-root) uid 0?

avikivity commented on test/manual/bti_cassandra_compatibility_test.py:
```diff
@@ -0,0 +1,621 @@
+            module_logger.info(f"Recreate keyspace.")
+            await bash(f"""podman exec {cluster_name}_{addrs[0]} cqlsh -e "DROP KEYSPACE IF EXISTS ks; CREATE KEYSPACE ks WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': 1}};" """)
+            module_logger.info(f"Delete potential leftovers from earlier tests.")
+            await bash(f"podman exec {cluster_name}_{addrs[0]} bash -c 'rm -rf /var/lib/cassandra/data/ks/*'")
```
Ah, but we'll defeat that soon.

michoecho commented on bin/clangd:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+cd "$(dirname "${BASH_SOURCE[0]}")/.."
```
For a moment I accidentally pushed a commit with a script specific to my local setup. Disregard this.

michoecho commented on test/manual/bti_cassandra_compatibility_test.py:
```diff
@@ -0,0 +1,621 @@
+        await bash(f"rm -rf {cassandra_dir}/upload && cp -r {workdir}/scylla {cassandra_dir}/upload")
+        with open(f"{cassandra_dir}/upload/schema_clone.cql", "w") as f:
+            f.write(clone_schema)
+        async with with_cluster("cassandra:latest", cassandra_dir) as (cluster_name, addrs, _):
```
Done.
